proposals/0540-jsem.rst

```diff
@@ -0,0 +1,384 @@
+
+Implementation of the protocol
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+We then propose that GHC and ``cabal``/``stack`` implement the GHC jobserver
+protocol, with GHC as a jobclient and ``cabal``/``stack` as the jobserver.
+
+As far as the GHC jobclient implementation is concerned, we have identified
+two salient points:
+
+A. There is a hidden cost in having a GHC program (such as ``ghc`` itself) run on
+   fewer CPU cores than its capabilities: the stop-the-world cost of garbage
+   collection becomes much more expensive. If we were to give GHC a fixed number
+   ``N`` of capabilities, then parallel garbage collections in that GHC would
+   recruit ``N`` OS threads. Instead we propose that GHC will adjust its number
+   of capabilities, via ``setNumCapabilities`` [8]_ , to the number of modules
+   it is compiling.
+   Note that GHC already calls ``setNumCapabilities`` with the argument to ``-j``.
+
+   We have concerns that rapidly adjusting the number of capabilities may have
+   adverse effects: we propose that GHC will limit the rate at which it can
+   release tokens to the semaphore so as to reduce this thrashing.
+   The exact mechanism of rate limiting is unspecified here.
+
+B. We believe that GHC invocations should not be too eager to release
+   semaphore tokens. This skews the balance in favour of in-unit parallelism
+   (1 unit with many capabilities), and against compiling many units in parallel
+   (many units with a single capability).
+   The justification is that the memory used by compiling units can be released
+   before starting another parallel process. Were GHC to release semaphore tokens
+   too eagerly, it could end up compiling a large number of packages in parallel
+   which each have a large loaded EPS. Combined, this will use a significant amount
+   of memory.
+
+   To achieve this, we propose a minimum delay between acquiring a resource and
+   releasing a resource. This allows us to prioritise completing a single large
+   package before moving on to other work. The exact mechanism through which this
+   is achieved is left unspecified in this proposal.
+
+Examples
+--------
+
+Let us explain how we envision ``cabal`` handle the following build plan, with
+8 capabilities.
+
+.. image:: jsem_modules_plain.svg
+  :width: 600
+  :alt: Same dependency graph as before: a big unit at the top and bottom,
+        and many single-module units in the middle.
+
+1. To start, ``cabal`` would create a semaphore ``⟨sem⟩`` with 8 available tokens.
+
+2. Next, we compile the ``Bot`` unit, which is a large unit, with many modules,
+   which sits at the bottom of the dependency graph and must thus be compiled
+   before anything else.
+
+   a. ``cabal`` acquires one token from the semaphore and spawns one
+      ``ghc --make -jsem ⟨sem⟩`` invocation.
+   b. This invocation of ``ghc`` notices it has a lot of work to do (many modules
+      to compile from the ``Bot`` unit), so it requests more resources from the
```

Comment:
Does GHC request a specific number of tokens, or just "more"?

Reply:
more = 1 per module it needs to compile.